feat(eslint-plugin): [strict-bool-expr] add fixes and suggestions

phaux opened the pull request:
Added automatic fixes and suggestions to make this rule more easy to use.

I added following fixes and suggestions for particular types in boolean context (also added to docs):

- `boolean` - Always allowed - no fix needed.
- `string` - (when `allowString` is `false`) - Provides following suggestions:
  - Change condition to check string's length (`str` → `str.length > 0`)
  - Change condition to check for empty string (`str` → `str !== ""`)
  - Explicitly cast value to a boolean (`str` → `Boolean(str)`)
- `number` - (when `allowNumber` is `false`) - Provides following suggestions:
  - Change condition to check for 0 (`num` → `num !== 0`)
  - Change condition to check for NaN (`num` → `!Number.isNaN(num)`)
  - Explicitly cast value to a boolean (`num` → `Boolean(num)`)
- `object | null | undefined` - (when `allowNullableObject` is `false`) - Provides **autofix**:
  - Change condition to check for null/undefined (`maybeObj` → `maybeObj != null`)
- `boolean | null | undefined` - Provides **autofix**:
  - Explicitly treat nullish value the same as false (`maybeBool` → `maybeBool ?? false`)
- `string | null | undefined` - Provides following suggestions:
  - Change condition to check for null/undefined (`maybeStr` → `maybeStr != null`)
  - Explicitly treat nullish value the same as an empty string (`maybeStr` → `maybeStr ?? ""`)
  - Explicitly cast value to a boolean (`maybeStr` → `Boolean(maybeStr)`)
- `number | null | undefined` - Provides following suggestions:
  - Change condition to check for null/undefined (`maybeNum` → `maybeNum != null`)
  - Explicitly treat nullish value the same as 0 (`maybeNum` → `maybeNum ?? 0`)
  - Explicitly cast value to a boolean (`maybeNum` → `Boolean(maybeNum)`)
- `any` and `unknown` - Provides following suggestions:
  - Explicitly cast value to a boolean (`value` → `Boolean(value)`)

Edit: updated order and messages to match the actual ones

--- 

I used the word "explicitly" where it doesn't change the runtime behavior, even when types are wrong.

The autofix for nullable object theoretically changes the behavior (e.g. if you somehow get empty string or 0 instead of null or object), but it's not enabled by default and there's only one obvious way of handling it if enabled. In practice it shouldn't make a difference when you compare object to null vs implicitly cast to boolean.

The second autofix for nullable boolean doesn't change the behavior and it's usually what the programmer wants. That's why I decided to make it automatic instead of suggestion.

I left some TODO comments about suggestions which are wrong for bigint (it's treated the same as number in the rule). They shouldn't cause a huge breakage when applied (in worst case TypeScript should report error after applying the suggestions). I think they can be fixed separately in a followup PR.

bradzacher commented on packages/eslint-plugin/src/rules/strict-boolean-expressions.ts:
```diff
@@ -93,6 +103,19 @@ export default util.createRule<Options, MessageId>({
       noStrictNullCheck:
         'This rule requires the `strictNullChecks` compiler option to be turned on to function correctly.',
+
+      conditionFixDefaultEmptyString:
+        'Explicitly treat nullish value the same as an empty string',
+      conditionFixDefaultZero: 'Explicitly treat nullish value the same as 0',
+      conditionFixCompareNullish:
+        'Change condition to check for null/undefined',
+      conditionFixCastBoolean: 'Explicitly cast value to a boolean',
+      conditionFixCompareStringLength:
+        "Change condition to check string's length",
+      conditionFixCompareEmptyString:
+        'Change condition to check for empty string',
+      conditionFixCompareZero: 'Change condition to check for 0',
+      conditionFixCompareNaN: 'Change condition to check for NaN',
```
suggestion for the suggestion text - show what the fixed code would look like to make it clearer.

EG:
```suggestion
      conditionFixDefaultEmptyString:
        'Explicitly treat nullish value the same as an empty string (`value ?? ''`)',
      conditionFixDefaultZero:
        'Explicitly treat nullish value the same as 0 (`value ?? 0`)',
      conditionFixCompareNullish:
        'Change condition to check for null/undefined (`value != null`)',
      conditionFixCastBoolean:
        'Explicitly cast value to a boolean (`Boolean(value)`)',
      conditionFixCompareStringLength:
        "Change condition to check string's length (`value.length !== 0`)",
      conditionFixCompareEmptyString:
        'Change condition to check for empty string (`value !== ""`)',
      conditionFixCompareZero:
        'Change condition to check for 0 (`value !== 0`)',
      conditionFixCompareNaN: 'Change condition to check for NaN (`!Number.isNaN(value)`)',
```

bradzacher reviewed (changes requested):
Logic itself LGTM.
One thing that needs to be addressed

----

I agree with all of the fixes except for the autofix for nullable boolean.

> Explicitly treat nullish value the same as false (maybeBool → maybeBool ?? false)

two reasons:
1) I think that it's a stylistic choice here and a lot of people won't like the autofix to `??`.
2) (the big one which blocks this) Not everyone is on TS 3.7, which would make this a breaking change because we'd be **auto**fixing to unsupported syntax and break their code.

Suggested solution:
make it a suggestion fixer with the following choices:
- `maybeBool ?? false`
- `maybeBool === true`

---

Side note for a (potential) future enhancement (not saying lets do it now - **god no**! - just saying someone might like it in the future).

Additional fixer for the verbose and explicit:
`x !== null && x !== undefined`
and the super verbose but super explicit and always safe:
`x !== null && typeof x !== 'undefined'`

phaux commented:
Done!

I changed the suggestions for nullable boolean as you suggested.

I also added a special case for number in boolean context when the number is `array.length` in which case it provides autofix instead of suggestion.

In addition, I moved `getWrappingFixer` to utils because it might be useful in other rules.

Then I also fixed a bug with `getWrappingFixer` so that it adds a semicolon if it detects that it inserted parens at the beginning of an ExpressionStatement.

bradzacher reviewed (approved):
thanks so much for your work on this!
This is a really valuable set of changes.
